lib/AST/ModuleNameLookup.cpp

```diff
@@ -90,20 +95,111 @@ static bool updateOverloadSet(NamedCanTypeSet &overloads,
+  ///
+  /// Both \p overloads and \p resolutionKind are updated based on the new
+  /// declarations.
+  ///
+  /// \returns true if any decls were added to \p results.
+  bool recordImportDecls(SmallVectorImpl<ValueDecl *> &results,
+                         ArrayRef<ValueDecl *> newDecls,
+                         OverloadSetTy &overloads,
+                         ResolutionKind &resolutionKind);
+
+  /// Given a list of imports and an access path to limit by, perform a lookup
+  /// into each of them and record the results in \p decls, filtering based on
+  /// the given resolution kind and existing overload set.
+  void collectLookupResultsFromImports(
+      ArrayRef<ModuleDecl::ImportedModule> imports,
+      ModuleDecl::AccessPathTy accessPath,
+      SmallVectorImpl<ValueDecl *> &decls,
+      const DeclContext *moduleScopeContext,
+      OverloadSetTy &overloads, ResolutionKind resolutionKind);
+
+  /// Performs a qualified lookup into the given module and, if necessary, its
+  /// reexports, observing proper shadowing rules.
+  void lookupInModule(ModuleDecl *module,  ModuleDecl::AccessPathTy accessPath,
+                      SmallVectorImpl<ValueDecl *> &decls,
+                      ResolutionKind resolutionKind,
+                      const DeclContext *moduleScopeContext,
+                      ArrayRef<ModuleDecl::ImportedModule> extraImports = {});
+public:
+  /// Performs a qualified lookup into the given module and, if necessary, its
+  /// reexports, observing proper shadowing rules.
+  static void lookupInModule(ModuleDecl *startModule,
+                             ModuleDecl::AccessPathTy accessPath,
+                             SmallVectorImpl<ValueDecl *> &decls,
+                             ResolutionKind resolutionKind,
+                             LazyResolver *typeResolver,
+                             const DeclContext *moduleScopeContext,
+                             ArrayRef<ModuleDecl::ImportedModule> extraImports,
+                             LookupCallbackTy callback) {
+    ModuleNameLookup lookup;
+    lookup.callback = callback;
+    lookup.typeResolver = typeResolver;
+    lookup.respectAccessControl =
+        !startModule->getASTContext().isAccessControlDisabled();
+
+    lookup.lookupInModule(startModule, accessPath, decls, resolutionKind,
+                          moduleScopeContext, extraImports);
+  }
+};
+} // end anonymous namespace
+
+template <typename OverloadSetTy>
+bool ModuleNameLookup<OverloadSetTy>::recordImportDecls(
+    SmallVectorImpl<ValueDecl *> &results,
+    ArrayRef<ValueDecl *> newDecls,
+    OverloadSetTy &overloads,
+    ResolutionKind &resolutionKind) {
+  const size_t originalSize = results.size();
+
   switch (resolutionKind) {
   case ResolutionKind::Overloadable: {
     // Add new decls if they provide a new overload. Note that the new decls
     // may be ambiguous with respect to each other, just not any existing decls.
```

Comment:
I realize that this wording is old, but does "any existing decls" mean any anywhere? Or just in one of these data structures here?